UI: Add obs_frontend_close_main_window()

BenJuan26 opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
<!--- Describe your changes in detail. -->
<!--- If this change includes UI elements, please include screenshots. -->

Adds the ability to close the main window from the frontend API. The behaviour is the same as clicking the Exit button in OBS Studio.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open GitHub Issue, or implements feature request -->
<!--- from the Ideas page, please link to the issue here. -->

This is useful for scripting, where OBS is mainly run from the system tray and controlled through the script, i.e. through obs-websocket. Currently, the only way to close OBS from a script is to have the OS terminate the process. This is obviously not ideal: the program doesn't get the chance to properly free up resources, and in the case of Windows, the icon hangs around the system tray until it is moused-over.

Link to feature suggestion: https://ideas.obsproject.com/posts/2225/allow-clean-shutdown-via-api

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment (hardware, OS version, etc.),-->
<!--- and the tests you ran, including how it may affect other areas of code. -->

OS: Windows
CPU: Intel i7 12700K
GPU: Gigabyte Nvidia GeForce RTX 3070 Ti 8 GB
RAM: 32GB

For the sake of simplicity, I hijacked the `StartStream` handler of obs-websocket to instead call `obs_frontend_close_main_window()`, then used OBS-web to trigger the handler. The OBS Studio window closed, and the log indicated that the shutdown process completed nominally, just as it would have if the Exit button was clicked.

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
- New feature (non-breaking change which adds functionality)
<!--- - Tweak (non-breaking change to improve existing functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
- Documentation (a change to documentation pages)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

PatTheMav commented:
Is this supposed to end the application? Because on macOS an application is not automatically terminated just because it has no active windows - that’s an opt-in behaviour. It could be that Qt behaves this way in our case, but that needs to be verified.

BenJuan26 commented:
For the use case that I'm targeting, yes, the intention is to end the application.

I just tested the release version of OBS on macOS, and indeed, clicking the Exit button and closing the window both end the application.

derrod commented:
> Is this supposed to end the application? Because on macOS an application is not automatically terminated just because it has no active windows - that’s an opt-in behaviour. It could be that Qt behaves this way in our case, but that needs to be verified.

Our main window's close event handler does some cleanup and ultimately calls `qApplication::quit()`

https://github.com/obsproject/obs-studio/blob/c289efd84a42887feba89c3579d39a9c264b749e/UI/window-basic-main.cpp#L5042

CodeYan01 commented on docs/sphinx/reference-frontend-api.rst:
```diff
@@ -242,6 +242,12 @@ Functions
 
+.. function:: void obs_frontend_close_main_window(void)
+
+   Closes the main window.
```
Perhaps some clarification here is in order. Like, would it prompt the user if there are active outputs, or ignore outputs.

BenJuan26 replied:
I added a line about active outputs. I didn't want to make things too wordy. I could have stated that behaviour is the same as the Exit button, but I think it's clearer to just state the behaviour.

CodeYan01 replied:
Currently looks good to me. However, I would prefer:
> Closes the main window**, as if the user clicked the Exit button**. If there are active outputs, the user will be prompted to confirm before closing.

Which may or may not be necessary, and I'll leave that opinion to you and the others.

BenJuan26 replied:
I'll leave it unresolved for now, so others can weigh in.

CodeYan01 reviewed (commented):
LGTM

Warchamp7 commented:
I'd like us to include a log line indicating when something has closed OBS this way.

asoulliereHT commented:
> I'd like us to include a log line indicating when something has closed OBS this way.
> 
> How does this interact with things that prevent closing the window such as running outputs?

"Closes the main window. If there are active outputs, the user will be prompted to confirm before closing."

BenJuan26 commented:
@Warchamp7 Log line added.

> How does this interact with things that prevent closing the window such as running outputs?

This was answered in the next comment. Let me know if you have any more questions/issues.

Warchamp7 reviewed (approved):
Fine from my end with the log line added

GeneralSpace commented:
@BenJuan26 @Warchamp7 this would actually really help with a project I'm working on. Any chance it could be considered for inclusion in 31.1 or another minor version soon?

BenJuan26 committed:
UI: Add obs_frontend_close_main_window()

Adds the ability to close the main window from the frontend API.
The behaviour is the same as clicking the Exit button in OBS Studio.

BenJuan26 commented:
@Warchamp7 @CodeYan01 This very simple change is nearly two years old and many parties have expressed interest in using it. Any idea whether there are plans to merge it anytime soon? Thanks.

RayneYoruka commented:
It seems I'm landing here too looking for a way to gracefully shutdown obs. I don't think taskill is the best way to do it. 

The log ends up being incomplete as it counts as a forced exit.

asoulliereHT commented:
I'm commenting here to bump.  I'm seeing wide support for this feature and no concerns on code.  🤷

RayneYoruka commented:
I'm returning back with a hackery solution yet this seems to be the best! 
NirCMD can gracefully exit obs without any issues, just make a bat file and automate in your favourite software and have it have "nircmd.exe closeprocess obs64.exe"

It exists as if you closed the X button to close!

BenJuan26 commented:
Closing as unnecessary. OBS now handles termination signals correctly.
